### ISSUE TITLE
Advantage against flash attention2

### ISSUE THREAD

**Issue by zhangmozhe** [google-deepmind/recurrentgemma#4]
Congratulations on your great work! I have a question regarding the performance against flash attention2. Since the flash attention 2 does not bring additional memory footprint besides KQV calculation and it's also pretty fast,  I want to know more about  the GPU memory utilization and throughput of the Griffin model against flash attention. In our initial experiment, flash attention is faster and more memory efficient. Is there any way to further optimize the Griffin model implementation to beat flash attention?

Your insight is well appreciated. Thanks a lot!

**Comment by botev** [comment]
There are a few things that are needed to be said here, so this might be a long reply. 

1. Yes flash attention does not bring additional memory, but the attention computation still has T^2 FLOPs. The way it works is through essentially turning the attention into two scans, one for the unnormalized values and one for the softmax normalizer (executed together). Assuming infinite hardware it will have linear time complexity. Otherwise, it will become compute bound for some value of T onwards. The RNN is always linear in time. So from those 2 components the RNN is always at least as fast. 
2. Most sampling is memory bound. The attention KV cache always grows unless you are using local attention only. As a result for large enough sequences that will start to dominate, which is why the throughout of attention starts to drop, even if you aren't saturating the device, since the bottleneck are memory transfers. 
3. The RecurrentGemma model and Griffin in general has different architecture hypers than Gemma. Eg it's a bit wider, a bit deeper, but has less expansion in the MLP. These things can effect full model runtime, even if individual components are faster. For comparing, you should try to design apples to apples comparison as much as possible. On this point note that our recurrent  block has more linear layers than the MQA block, so for very wide models this might dominate the computation.
4. When comparing such things for individual parts it is important to have a well optimised implementation for both. Currently our code is only optimized in Jax and for TPUs specifically. We generally don't have expertise to do this for GPUs or PyTorch. Our PyTorch code is more for a reference, but is definitely not optimized well. We are happy to take PRs in this direction, but probably won't spend time ourselves on it. 
5. Finally, could you share the exact settings/things that you are comparing, so we might be able to potentially give a bit more detailed advice. At the moment the question is a bit too general to be able to give you pointers.

**Comment by Nush395** [comment]
Hi @zhangmozhe, hope that your question is answered/resolved, can we close this issue now?

**Comment by zhangmozhe** [comment]
Sure, thanks a lot for your reply. Great work!
